### PR TITLE
Only convert inputs to FP16 when FP16 stage is used

### LINKED ISSUE #336: Only convert inputs to FP16 when FP16 stage is used
## Issue

`src/onnxflow/common/tensor_helpers.py:30: save_inputs()` converts inputs to FP16 even when the FP16 is not part of the user-provided custom sequence.

## Task

Ensure that inputs are only converted to FP16 when the FP16 stage is used.

### PR DESCRIPTION
**Pull request by danielholanda**
Closes #336

## Description

This PR ensures that 
* inputs are ONLY converted to FP16 if we are running the FP16 stage.
* `expected_input_dtypes` are updated when the FP16 stage is used.

## Testing

A test was added to `build_model.py`. You can also test the changes using the script below:

```
# labels: name::linear author::selftest test_group::selftest
import torch
import numpy as np
import os
from groqflow import groqit
import onnxflow.justbuildit.export as export
import onnxflow.justbuildit.stage as stage
from onnxflow.justbuildit.buildit import build_model

torch.manual_seed(0)
cache_location = ".cache/onnxflow_test_cache"


custom_sequence_fp32 = stage.Sequence(
    "custom_sequence_fp32",
    "Building Pytorch Model without fp16 conversion",
    [
        export.ExportPytorchModel(),
        export.OptimizeOnnxModel(),
    ],
    enable_model_validation=True,
)

custom_sequence_fp16 = stage.Sequence(
    "custom_sequence_fp16",
    "Building Pytorch Model with fp16 conversion",
    [
        export.ExportPytorchModel(),
        export.OptimizeOnnxModel(),
        export.ConvertOnnxToFp16(),
    ],
    enable_model_validation=True,
)

torch.manual_seed(0)

# Define model class
class SmallModel(torch.nn.Module):
    def __init__(self, input_size, output_size):
        super(SmallModel, self).__init__()
        self.fc = torch.nn.Linear(input_size, output_size)

    def forward(self, x):
        output = self.fc(x)
        return output


# Instantiate model and generate inputs
input_size = 10
output_size = 5
pytorch_model = SmallModel(input_size, output_size)
inputs = {"x": torch.rand(input_size)}

# Build model using fp32 inputs
build_name = "custom_sequence_fp32"
omodel = build_model(
    pytorch_model,
    inputs,
    build_name=build_name,
    rebuild="always",
    monitor=False,
    cache_dir=cache_location,
    sequence=custom_sequence_fp32,
)

inputs_path = os.path.join(cache_location, build_name, "inputs.npy")
assert np.load(inputs_path, allow_pickle=True)[0]["x"].dtype == np.float32

# Build model using fp16 inputs
build_name = "custom_sequence_fp16"
omodel = build_model(
    pytorch_model,
    inputs,
    build_name="custom_sequence_fp16",
    rebuild="always",
    monitor=False,
    cache_dir=cache_location,
    sequence=custom_sequence_fp16,
)

inputs_path = os.path.join(cache_location, build_name, "inputs.npy")
assert np.load(inputs_path, allow_pickle=True)[0]["x"].dtype == np.float16
```

### CODE REVIEW

File: src/onnxflow/justbuildit/export.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import copy
 from typing import Union
 import torch
+import numpy as np
```
Comment (jeremyfowers) [inline src/onnxflow/justbuildit/export.py:9]: Please rev the onnxflow and mlagility patch version numbers so that we can release a package.

**Review by jeremyfowers — APPROVED** [review]
This is great! It's funny that we already had the `downcast` arg but just never used it. 

I approve, but I suggest that you ask @vgodsoe-groq to test the changes within GroqFlow on a GroqNode before merging (just to avoid any potential back and forth additional PRs).

File: src/onnxflow/justbuildit/export.py (hunk 1 above)
Comment (danielholanda) [reply src/onnxflow/justbuildit/export.py:9]: Updated both version numbers!

**Comment by danielholanda** [comment]
I have a decision to make about this PR and my Magic 8 ball seems to be broken. Which solution is the most sound here?

**Issue:**
We are now ONLY converting the inputs to FP16 when the FP16 stage is run. This means that models that use the torch-eager/torch-compiled runtimes will now run in FP32 instead of FP16.
A consequence of this is that benchmarking a model using both ort and torch-eager (assuming the default sequences) will cause results to be overwritten due to the mismatched data types.

**Potential solutions:**
(1) Convert the inputs to FP16 also when torch-eager/torch-compiled are used
(2) Only check for input shapes, but not data types before building a model. Note: `*_state.yaml` will only keep track of the `expected_input_dtype` of the last model built if the code is not meaningfully refactored.
(3) Change the hashing logic to also include input dtypes

**Review by vgodsoe-groq — APPROVED** [review]

**Pull request merged**